benchclients/python/benchclients/http.py

```diff
@@ -0,0 +1,414 @@
+        # self._login_or_raise() has to persist its authentication state here,
+        # via e.g. cookies.
+        self.session = requests.Session()
+
+    @property
+    @abstractmethod
+    def _base_url(self) -> str:
+        """
+        Base URL. Child class must implement this. I would have picked an
+        actual abstract property (instead of a method), but this here seems
+        to be stricter.
+        """
+
+    @abstractmethod
+    def _login_or_raise(self) -> None:
+        """
+        Perform login.
+
+        Is expected to set self.session to an instance of requests.Session.
+
+        Persist authentication state in self.session.
+
+        Raise RetryingHTTPClientBadCredentials or RetryingHTTPClientLoginError.
+
+        The implementation should use
+
+                _make_request_retry_until_deadline()
+
+        for doing the login request (with persistent retrying).
+        """
+
+    def _abs_url_from_path(self, path) -> str:
+        assert path.startswith("/")
+        return self._base_url.rstrip("/") + path
+
+    def get(self, path: str, params: Optional[dict] = None) -> Union[Dict, List]:
+        """
+        Make GET request. Expect response with status code 200, expect a JSON
+        document in the response body.
+
+        Return the deserialized JSON document or raise an exception.
+
+        `params` can be used to pass URL query parameters.
+        """
+        resp = self._make_request(
+            "GET", self._abs_url_from_path(path), 200, params=params
+        )
+        return resp.json()
+
+    def put(self, path: str, json: Dict) -> Optional[Union[Dict, List]]:
+        """
+        Make PUT request. Send a JSON document in the request body. Expect
+        response with status code 201.
+
+        If the response has no body then return `None`.
+
+        If the response has a body of non-zero length then expect a JSON
+        document.
+
+        Return the deserialized JSON document or raise an exception.
+
+        Interface inherited from previous lib.
+        """
+        json = json or {}
+
+        resp = self._make_request("PUT", self._abs_url_from_path(path), 201, json=json)
+
+        if resp.content:
+            return resp.json()
+
+        return None
+
+    def post(
+        self, path: str, json: Optional[dict] = None
+    ) -> Optional[Union[Dict, List]]:
+        """
+        Make POST request. Send a JSON document in the request body. Expect
+        response with status code 201.
+
+        If the response has no body then return `None`.
+
+        If the response has a body of non-zero length then expect a JSON
+        document.
+
+        Return the deserialized JSON document or raise an exception.
+
+        Interface inherited from previous lib.
+        """
+        json = json or {}
+
+        if json:
+            log.debug("POST request JSON body:\n%s", jsondumps(json, indent=2))
+        else:
+            log.debug("POST request without body. Hm.")
```

Comment:
Funny enough it is possible to use the following POST endpoint with no body:
https://github.com/conbench/conbench/blob/abab673c9cfb02396a1fa8e5ab6fe7b4fa5d499b/benchalerts/benchalerts/integrations/github.py#L98
See: [GitHub API docs](https://docs.github.com/en/rest/apps/apps?apiVersion=2022-11-28#create-an-installation-access-token-for-an-app)